### ISSUE TITLE
Confusion Matrix weird results?

### ISSUE THREAD

**Issue by a-esp-1** [ultralytics/yolov5#1665]
## ❔Question
Hi,

I've seen you have update the repo (#1474) and add the Confusion Matrix. When I run it with my own dataset I get this Confusion Matrix:
![confusion_matrix](https://user-images.githubusercontent.com/73425046/101797425-519eb780-3b0a-11eb-9fd9-177a28bd3552.png)

My mAP is 0.84 for this dataset, and for example for Airplane class, AP = 0.95.
![ap](https://user-images.githubusercontent.com/73425046/101798045-f1f4dc00-3b0a-11eb-839d-750583a53ea5.PNG)

So I can't understand how is possible that in the Confusion Matrix [row=Airplane, column=Airplane] = 0.26, [row=Airplane, column=FN] = 0 and [row=FP, column=Airplane] = 0.73.

As I can understand this means that just 26% of detections are correct (TP), but it's not possible because AP(Airplane) = 0.95 and when I run detections Airplanes are well detected.

## Additional context
I'm getting this weird result not just with this dataset, if not with other 2.

**Comment by glenn-jocher** [comment]
@a-esp-1 in object detection the confusion matrix is unfortunately somewhat uninformative, as the majority of mistakes will be with the background class rather than another class.

Confusion matrix implementation is correct. Results will vary based on the confidence threshold however. As this threshold trends to 1.0 background FPs will trend towards zero. See https://github.com/kaanakan/object_detection_confusion_matrix/issues/7 for a more detailed discussion of this.

**Comment by glenn-jocher** [comment]
@a-esp-1 the confusion matrix conf threshold is here. The default value is 0.25, which is aligned with the default inference setting in detect.py and the model autoshape wrapper.
https://github.com/ultralytics/yolov5/blob/b2bef8f6d8e4c008bae72c211a186d75732fc213/utils/metrics.py#L107-L109

So theoretically the confusion matrix result should align with the default detect.py result. Are you seeing something different?

**Comment by a-esp-1** [comment]
I see. I might change conf threshold because in my detection I used 0.5, and here it is set to 0.25. I'm not complety sure but I think iou_thres was set to 0.65 (but I'm not 100% sure).

Thank you.

**Comment by a-esp-1** [comment]
@glenn-jocher I revised my data, and I've runned a test again. This are the results
![results](https://user-images.githubusercontent.com/73425046/101828432-d5b86580-3b31-11eb-9150-458680108f7a.png)


As you can sea mAP = 0.84 and Airplane AP = 0.96. This matches with what I can see in my detections. But this doesn't match with the ConfusionMatrix obtainned.

![confusion_matrix](https://user-images.githubusercontent.com/73425046/101827842-1663af00-3b31-11eb-9400-4ca2e5920c68.png)

If Airplane AP is about 0.95 it can't have 73% of FP.

For example, airport has AP = 0.92 and the ConfusionMaxtrix shows 94% of TP. I think this can be correct.

**Comment by glenn-jocher** [comment]
@a-esp-1 the confusion matrix is saying that across your 5000 test images, 1 out of every 4 plane detections is correct, the other 3 are FPs at your selected NMS thresholds.

You can not draw any comparison with mAP, which is simply the area under the PR curve. A good mAP produced by a low confidence threshold will necessarily contain thousands of FPs, pushed to the bottom right of the curve as recall trends to 1 and precision to zero.

**Comment by a-esp-1** [comment]
Ok thank you.

**Comment by glenn-jocher** [comment]
I think what’s happening is you are seeing a few TP airplane predictions in a small number of images, but you are not seeing the FPs in the other 5k images because they are more rare and harder to spot.

**Comment by glenn-jocher** [comment]
Maybe if we had TP FP columns this would be easier to see. This might be a useful addition to test.py. I don’t have time myself, but if you’d like to submit a PR for this it may help other users as well.

**Comment by a-esp-1** [comment]
@glenn-jocher  I'm looking it but I have a doubt. Is this line suppose to be TPs?
https://github.com/ultralytics/yolov5/blob/94a7f55c4e5cca3dfe4de0bd0793173d5b152ec5/utils/metrics.py#L146

Because I tried to change this line
https://github.com/ultralytics/yolov5/blob/94a7f55c4e5cca3dfe4de0bd0793173d5b152ec5/utils/metrics.py#L162

to this

array = self.matrix

So I think this should mean to write TP in main diagonal. The results obtained are this:
![confusion_matrix](https://user-images.githubusercontent.com/73425046/101905065-21f9b900-3bb7-11eb-8a73-49ea0621b658.png)


This is a bit weid because the number of airplanes is 2746 (I mean objects, no images). So if airplane AP = 0.95, R = 0.94 and P = 0.87, this should mean that the number of airplane TPs should be > 316.

In the other hand, when I see that the number of airplane FN is 2430 this makes my mind explote, because 316 (TPs?) + 2746 (FN) = 2746 (groundtruth).



For all of this, I think it might be a problem in TP calculation.




You can see the number of objects of each class in the red column.
![Captura](https://user-images.githubusercontent.com/73425046/101905150-448bd200-3bb7-11eb-862b-f76782facfad.PNG)

**Comment by glenn-jocher** [comment]
@a-esp-1 if you've got questions about a specific function, you may want to provide results on a common dataset like COCO128.

**Comment by a-esp-1** [comment]
Ok @glenn-jocher, if I have time I'll try later.

But is this line suppose to calculate TPs?
https://github.com/ultralytics/yolov5/blob/94a7f55c4e5cca3dfe4de0bd0793173d5b152ec5/utils/metrics.py#L146

**Comment by glenn-jocher** [comment]
@a-esp-1 I think so yes, but I'm not the author of this code. The original author and code are from https://github.com/kaanakan/object_detection_confusion_matrix

**Comment by wonh** [comment]
> @a-esp-1 the confusion matrix is saying that across your 5000 test images, 1 out of every 4 plane detections is correct, the other 3 are FPs at your selected NMS thresholds.
> 
> You can not draw any comparison with mAP, which is simply the area under the PR curve. A good mAP produced by a low confidence threshold will necessarily contain thousands of FPs, pushed to the bottom right of the curve as recall trends to 1 and precision to zero.

I quite don't agree with you. A good mAP means high precision under high recall, and it contens few FPs.

**Comment by glenn-jocher** [comment]
@wonh if mAP is measured at the ideal confidence threshold of 0.0, it will by definition contain hundreds or even thousands of times more FPs than TPs, therefore my above statement is correct. It is the location of the detections on the curve that influences mAP, not the quantities or ratios.

**Comment by wonh** [comment]
@glenn-jocher if confidence threshold is 0.0, there will be hundreds or even thousands of times more FPs than TPs. So the recall will improve but the precision will goes down. How can you get a good mAP?  If mAP is 1.0, there will be no FPs .

**Comment by a-esp-1** [comment]
@wonh mAP is the area under precision-recall curve. For generating that curve confidence changes from 1 to 0. This means that in a hypothetically perfect scenario, the curve would start in (1,1) and end up in (0,0), so the area under this curve would be 1 (mAP=1).

**Comment by wonh** [comment]
> @wonh mAP is the area under precision-recall curve. For generating that curve confidence changes from 1 to 0. This means that in a hypothetically perfect scenario, the curve would start in (1,1) and end up in (0,0), so the area under this curve would be 1 (mAP=1).

Thanks for your reply.I konw the meaning of mAP. My point is that mAP is the performance metric of the model. A good mAP conflict with thousands of times more FPs than TPs.

**Comment by glenn-jocher** [comment]
@wonh a good mAP can produce thousands of times more FPs than TPs. This is the definition of mAP. 

The location on the curve is what matters, not the ratio.

**Comment by nlm-yuh5** [comment]
hi @glenn-jocher , I have a question about the "confusion_matrix.png" generated during "train.py". Does it use the data from the "valid" folder or "test" folder?

**Comment by glenn-jocher** [comment]
@nlm-yuh5 test.py evaluates label metrics on 'train', 'val', or 'test' sets according to the --task argument (default 'val'):
https://github.com/ultralytics/yolov5/blob/52c05707d6ef02ed34c154b84e0f6a5d1f36ae37/test.py#L297

**Comment by a227799770055** [comment]
Hi, I have the question about the recall.
The below block show that my recall reach 0.918 at this test.

![截圖 2021-11-10 上午11 42 47](https://user-images.githubusercontent.com/50135574/141046397-c0bd2116-2c21-46c0-af36-1a33950e466b.png)

However, the confusion matrix show my TP = 0.97 and FN = 0.03, then the recall calculate with this two numbers is 0.97.
Why it's different result between the block show in testing and the confusing matrix generate by this testing?

![confusion_matrix](https://user-images.githubusercontent.com/50135574/141046368-6b0ad857-7097-4466-91d1-a0fb362aebb9.png)

**Comment by glenn-jocher** [comment]
@a227799770055 displayed P and R values are computed at maximum F1. F1 is evaluated over a thousand confidence values from 0 to 1 and the maximum point is selected.
https://github.com/ultralytics/yolov5/blob/7ebb5e5da673350d4c168cf60d01986a5e0f00cc/utils/metrics.py#L72-L84

EDIT: See F1_curve.png for this max point.
![image](https://user-images.githubusercontent.com/26833433/141107959-756e17e8-4a26-40dd-9107-11de9b21e1c5.png)

Confusion matrix results are computed at a fixed conf=0.25:
https://github.com/ultralytics/yolov5/blob/7ebb5e5da673350d4c168cf60d01986a5e0f00cc/utils/metrics.py#L114-L116

**Comment by a227799770055** [comment]
@glenn-jocher Tanks a lot!!

**Comment by glenn-jocher** [comment]
@a227799770055 I understand the confusion though. Maybe we should consolidate a single confidence to use for both of these metrics so they align and the users are less confused.

Typically max F1 will be your best balance between high P and high R, so for many use cases this is the optimal deployment confidence threshold.

**Comment by kashishgoyal31** [comment]
Please explain the confusion matrix w.r.t below situation. I am not able to interpret it. I have attached the training results. Please explain the results.

Regards
Kashish Goyal
![confusion_matrix](https://user-images.githubusercontent.com/89016389/165912050-28269585-405f-4aa4-b1ef-33b16be0c667.png)
![Screenshot 2022-04-29 at 2 03 14 PM](https://user-images.githubusercontent.com/89016389/165912101-f84a3464-3068-4298-8c25-001558eb7a64.png)

**Comment by glenn-jocher** [comment]
@kashishgoyal31 confusion matrix shows good results.

**Comment by kashishgoyal31** [comment]
Sir could you please help me interpret it w.r.t my problem. I have attached
the screenshot for your reference. How will I explain it.

**Comment by yizweithree** [comment]
Hi, I have the question about the confusion matrix. Why is the sum of the values of each column of the confusion matrix less than 1 or more than 1, such as 0.98, 1.02, etc.? Thank you.

**Comment by glenn-jocher** [comment]
@yizweithree 👋 hi, thanks for letting us know about this possible problem with YOLOv5 🚀. We've created a few short guidelines below to help users provide what we need in order to start investigating a possible problem.

## How to create a Minimal, Reproducible Example

When asking a question, people will be better able to provide help if you provide **code** that they can easily understand and use to **reproduce** the problem. This is referred to by community members as creating a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/). Your code that reproduces the problem should be:

- ✅ **Minimal** – Use as little code as possible to produce the problem
- ✅ **Complete** – Provide all parts someone else needs to reproduce the problem
- ✅ **Reproducible** – Test the code you're about to provide to make sure it reproduces the problem

For [Ultralytics](https://ultralytics.com/) to provide assistance your code should also be:

- ✅ **Current** – Verify that your code is up-to-date with GitHub [master](https://github.com/ultralytics/yolov5/tree/master), and if necessary `git pull` or `git clone` a new copy to ensure your problem has not already been solved in [master](https://github.com/ultralytics/yolov5/tree/master).
- ✅ **Unmodified** – Your problem must be reproducible using official YOLOv5 code without changes. [Ultralytics](https://ultralytics.com/) does not provide support for custom code ⚠️.

If you believe your problem meets all the above criteria, please close this issue and raise a new one using the 🐛 **Bug Report** [template](https://github.com/ultralytics/yolov5/issues/new/choose) with a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) to help us better understand and diagnose your problem. 

Thank you! 😃

**Comment by marziyemahmoudifar** [comment]
I used the yolov5 network for object detection on my dataset which has only one class. But I do not understand the confusion matrix. Why FP is one and TN is zero?
![confusion_matrix](https://user-images.githubusercontent.com/77172527/171788396-8e45fea4-5161-437d-a405-9870cfac7508.png)

**Comment by glenn-jocher** [comment]
@marziyemahmoudifar columns are normalized

**Comment by marziyemahmoudifar** [comment]
> @marziyemahmoudifar columns are normalized

Well, my question is that the main diameter of the matrix should have the largest values ​​and the sub-diameter less, but this is not the case here !!

**Comment by glenn-jocher** [comment]
Your assumption is incorrect, the matrix may have any values depending on validation results, the only constraint is normalized columns as I explained above.

**Comment by marziyemahmoudifar** [comment]
> Your assumption is incorrect, the matrix may have any values depending on validation results, the only constraint is normalized columns as I explained above.

Thank you so much...

**Comment by jjjonathan14** [comment]
![index](https://user-images.githubusercontent.com/104885230/185398394-2b335a5c-0d90-4dcd-8689-09af44e002f8.png)

We have trained with two classes(free hand, hand with object) model and have enough background. But it doesnot predict any background. Do we need to label the background as well>

**Comment by glenn-jocher** [comment]
@jjjonathan14 only labelled data is predicted in images, background is not predicted.

**Comment by jjjonathan96** [comment]
Thanks for the reply.

**Comment by jjjonathan96** [comment]
Thanks for the reply.

**Comment by karantai** [comment]
@glenn-jocher So, if we change the conf thresh from ConfusionMatrix class a new conf matrix will appear with different TP FP AND FN and in order to compute the TRUE recall and precision of our dataset we need to compute them on our own? Because the terminal precision and recall is always the same, that is the best precision and recall during training set using a specific validation set. Am I missing something or am I correct?

**Comment by ghost** [comment]
@glenn-jocher  
![image](https://user-images.githubusercontent.com/44346562/211882354-e8568218-e0d5-4f0a-ba40-be2529e95812.png)
hello glenn, I have trained yolov5 with two classes(bad stacks, good stacks) background FP = .40, .60 what does that mean?  and how background FP and FN are calculated.

**Comment by enesayan** [comment]
> @glenn-jocher I'm looking it but I have a doubt. Is this line suppose to be TPs?
> 
> https://github.com/ultralytics/yolov5/blob/94a7f55c4e5cca3dfe4de0bd0793173d5b152ec5/utils/metrics.py#L146
> 
> Because I tried to change this line
> 
> https://github.com/ultralytics/yolov5/blob/94a7f55c4e5cca3dfe4de0bd0793173d5b152ec5/utils/metrics.py#L162
> 
> to this
> 
> array = self.matrix
> 
> So I think this should mean to write TP in main diagonal. The results obtained are this: ![confusion_matrix](https://user-images.githubusercontent.com/73425046/101905065-21f9b900-3bb7-11eb-8a73-49ea0621b658.png)
> 
> This is a bit weid because the number of airplanes is 2746 (I mean objects, no images). So if airplane AP = 0.95, R = 0.94 and P = 0.87, this should mean that the number of airplane TPs should be > 316.
> 
> In the other hand, when I see that the number of airplane FN is 2430 this makes my mind explote, because 316 (TPs?) + 2746 (FN) = 2746 (groundtruth).
> 
> For all of this, I think it might be a problem in TP calculation.
> 
> You can see the number of objects of each class in the red column. ![Captura](https://user-images.githubusercontent.com/73425046/101905150-448bd200-3bb7-11eb-862b-f76782facfad.PNG)

Did you solve the problem? Because, I have the same problems, test outputs, and confusion matrix outputs are different. For example for a specific class iou:60, conf=0.25 test values TP : 221, FN:60, FP:85 but in confusion matrix : TP : 239, FN:42, FP:92 with same iou and conf values. I don't understand what am I missing?

**Comment by glenn-jocher** [comment]
@enesayan thanks for reaching out. The confusion matrix is calculated based on the validation set predictions and ground truth labels. If you're getting different results between the test outputs and the confusion matrix, it could be due to a few reasons. It's possible that the test set contains different images with different set of ground truth boxes as compared to the validation set. Additionally, the confusion matrix is calculated based on the confidence threshold which is set to 0.5 by default, whereas, the test results shown are based on the confidence and/or IOU threshold that you set in your command to run the test. 

Regarding the specific example you provided, the differences in the number of TP, FP, and FN between the test outputs and the confusion matrix could be due to different threshold settings. The test output is probably based on a different threshold setting than the confusion matrix. As a result, the number of detections and misclassifications could vary.

In summary, the confusion matrix should be considered as the primary evaluation metric for your model, and it is calculated based on a fixed threshold of 0.5 by default. If you want to evaluate your model with different threshold settings, you can run the test set and look at the precision and recall values for each class. Hopefully, this clears up some of your doubts. If you have any other questions, feel free to ask.

**Comment by enesayan** [comment]
> @enesayan thanks for reaching out. The confusion matrix is calculated based on the validation set predictions and ground truth labels. If you're getting different results between the test outputs and the confusion matrix, it could be due to a few reasons. It's possible that the test set contains different images with different set of ground truth boxes as compared to the validation set. Additionally, the confusion matrix is calculated based on the confidence threshold which is set to 0.5 by default, whereas, the test results shown are based on the confidence and/or IOU threshold that you set in your command to run the test.
> 
> Regarding the specific example you provided, the differences in the number of TP, FP, and FN between the test outputs and the confusion matrix could be due to different threshold settings. The test output is probably based on a different threshold setting than the confusion matrix. As a result, the number of detections and misclassifications could vary.
> 
> In summary, the confusion matrix should be considered as the primary evaluation metric for your model, and it is calculated based on a fixed threshold of 0.5 by default. If you want to evaluate your model with different threshold settings, you can run the test set and look at the precision and recall values for each class. Hopefully, this clears up some of your doubts. If you have any other questions, feel free to ask.

Thank you for your quick response, firstly I am using the validation set as a test set also. There are the same images and labels in the test and validation data. Second I am using the same confidence and iou thresh in tests and in creating confusion matrix. I updated the below code in the confusion matrix and tested the model I set the same parameters but the results were different. 

iou:60, conf=0.25 same in test and confusion matrix. Test and validation the same data.

tets results=TP : 221, FN:60, FP:85
confusion matrix results=TP : 239, FN:42, FP:92

python test.py --weights runs/train/AP/yolov7/weights/last.pt --data data/data.yaml --name results/exp --batch-size 32 --task test --img-size 640  --device 1 --iou-thres 0.60 --conf-thres 0.25

"In summary, the confusion matrix should be considered as the primary evaluation metric for your model, and it is calculated 
based on a fixed threshold of 0.5 by default."

metrics.py
class ConfusionMatrix:
    # Updated version of https://github.com/kaanakan/object_detection_confusion_matrix
    def __init__(self, nc, conf=0.25, iou_thres=0.60):

**Comment by glenn-jocher** [comment]
@enesayan I apologize for any confusion caused. Based on the updated code you provided, I can see that the IOU and confidence threshold settings are consistent between your test and confusion matrix. However, the calculation of TPs, FPs, and FNs in the confusion matrix is based on the detected classes and their corresponding ground truth labels. Therefore, it's possible that some of the detections in the test set were misclassified and did not result in a true positive, which could explain the differences in the confusion matrix.

I would recommend double-checking your ground truth labels and test set detections to ensure that they are accurately labelled and that there are no discrepancies between them. Additionally, since you are using the validation set as your test set, it might be worth running your tests on a new set of images to ensure that you are not overfitting to your validation set.

If you continue to have issues or have any further questions, please don't hesitate to ask.

**Comment by enesayan** [comment]
Sorry, but I don't understand exactly what you mean. Isn't the confusion matrix obtained according to the results obtained when I call the test.py code anyway? Then why is there a difference between the test result written to the terminal and the confusion matrix? I do not understand this. There is missing something. I know, I will use a separate test set. But, I need to make sure that the model gives correct results with the validation data. 


python test.py --weights runs/train/AP/yolov7/weights/last.pt --data data/data.yaml --name results/exp --batch-size 32 --task test --img-size 640 --device 1 --iou-thres 0.60 --conf-thres 0.25

total images: 85 total labels for Apple 281
The terminal output:
Precision:0.72 Recall 0.79 
According to the formula 
TP =Recallxtotal labels =0.79x281=222
FP=(TP/Precision)-TP=(222/0.72)-222=86
FN=281-222=59

But in the confusion matrix 
TP: 239
FP:91
FN: 42

Also Confusion matrix class is calling in the test.py so the results should be the same but they are not. Do you have any idea?

**Comment by glenn-jocher** [comment]
@enesayan hello, thank you for your inquiry. The confusion matrix is calculated based on the detection outputs of your model and compared against the ground truth labels, whereas the test results shown in the terminal are based on the specified confidence and IOU threshold settings. Therefore, the confusion matrix and the terminal results may have differences in their outputs based on the different settings used to obtain them. 

To ensure that your model is correctly detecting objects in the validation data, it is advisable to use a separate test set with different images than those used in the validation data. This way, you can be sure that your model is not overfitting to the validation data. 

Regarding your specific example, it is possible that some of the detections in the test set were misclassified or overlapping, which could result in different TPs, FPs, and FNs being calculated between the confusion matrix and the terminal results. I would recommend further analyzing your test set and ground truth labels to ensure their accuracy and consistency.

If you have any further questions or concerns, please don't hesitate to ask.

**Comment by vishwa15-lab** [comment]
@enesayan Yes, the no. of output bounding boxes during test.y is not matching with no. of objects in confusion matrix.
@glenn-jocher any advice. Thanks!

**Comment by glenn-jocher** [comment]
@vishwa15 the number of output bounding boxes during the test and the number of objects in the confusion matrix not matching could be due to a few reasons. One possibility is that the test set and the validation set used to calculate the confusion matrix have different images or ground truth labels. It's important to ensure that both sets are consistent.

Another potential explanation could be that the thresholds used for the test and the confusion matrix are different. The test results displayed in the console are based on the specified confidence and IOU thresholds, whereas the confusion matrix is calculated using a fixed threshold of 0.5 by default. This difference in thresholds could result in variations between the two sets of results.

To further investigate the issue, I would recommend examining the specific images and labels in both the test set and the validation set. Double-checking their consistency and ensuring that the same thresholds are being used might help identify any discrepancies.

If the issue persists or if you have any other questions, please let us know.

**Comment by vishwa15-lab** [comment]
@glenn-jocher Thanks. To summarise:
- All predictions from the model with confidence score > specified & iou(NMS) > specified are put into txt file & used for calculating different MAP. This can be tested by lowering the any of the two thresholds. The Recall will increase & Precision will decrease.
- But for generating the confusion matrix, the proposals with confidence score > 0.25 and iou > 0.45 are only considered & this can't be controlled unless the code is modified inside. This is done in utils/metrics.py (line 131, 136)
- During inference, we can control only confidence score & can't control iou because there is no ground truth. Although iou for NMS is set to 0.45 by default & can't be controlled unless code is modified inside.

**Comment by glenn-jocher** [comment]
@vishwa15 the predictions from the model are saved in a txt file and used to calculate different mean average precision (MAP) metrics when the confidence scores and IOU (Intersection over Union) thresholds are met. Lowering either of these thresholds will increase recall but decrease precision.

On the other hand, when generating the confusion matrix, only proposals with confidence scores > 0.25 and IOU > 0.45 are considered. Modifying this behavior would require making changes to the code in utils/metrics.py (line 131, 136).

During inference, we can only control the confidence scores, as there are no ground truth values to calculate IOU. By default, the IOU threshold for Non-Maximum Suppression (NMS) is set to 0.45 and cannot be controlled without modifying the code.

If you have any further questions or need additional clarification, please let us know.

**Comment by rahafAlerwi** [comment]
> @nlm-yuh5 test.py evaluates label metrics on 'train', 'val', or 'test' sets according to the --task argument (default 'val'):
> 
> https://github.com/ultralytics/yolov5/blob/52c05707d6ef02ed34c154b84e0f6a5d1f36ae37/test.py#L297

while i'm training my custom dataset on yolov8 i noticed that the confusion matrix for the validation and training are the same .. how is that?     
this is my yaml 
![image](https://github.com/ultralytics/yolov5/assets/127831012/02ea493d-bfff-4cad-9eac-c647478d7810)
and this is the training
![image](https://github.com/ultralytics/yolov5/assets/127831012/d7f89176-b214-4fa1-bd12-1a4ef778201d)
this the confusion matrix for the training 
![image](https://github.com/ultralytics/yolov5/assets/127831012/8506e637-7267-469a-b286-c1294fa071fd)
and this the validation
![image](https://github.com/ultralytics/yolov5/assets/127831012/dd7c26d0-dd04-4b38-bc10-513cc683eca4)
and the confusion matrix for validation
![image](https://github.com/ultralytics/yolov5/assets/127831012/999b9223-59ef-4e1c-82db-d6c19a7914ae)

i you could help me i will be thankful

**Comment by abetancordelrosario** [comment]
> Sorry, but I don't understand exactly what you mean. Isn't the confusion matrix obtained according to the results obtained when I call the test.py code anyway? Then why is there a difference between the test result written to the terminal and the confusion matrix? I do not understand this. There is missing something. I know, I will use a separate test set. But, I need to make sure that the model gives correct results with the validation data.
> 
> python test.py --weights runs/train/AP/yolov7/weights/last.pt --data data/data.yaml --name results/exp --batch-size 32 --task test --img-size 640 --device 1 --iou-thres 0.60 --conf-thres 0.25
> 
> total images: 85 total labels for Apple 281 The terminal output: Precision:0.72 Recall 0.79 According to the formula TP =Recallxtotal labels =0.79x281=222 FP=(TP/Precision)-TP=(222/0.72)-222=86 FN=281-222=59
> 
> But in the confusion matrix TP: 239 FP:91 FN: 42
> 
> Also Confusion matrix class is calling in the test.py so the results should be the same but they are not. Do you have any idea?

@enesayan  Did you find a solution to this problem?

**Comment by froth0203** [comment]
@abetancordelrosario @enesayan @glenn-jocher: There is still the same problem using yolov9. Setting the conf and iou thresh as same for both validation and confusion matrix does not work...

One solution could be using the predict.py instead of val.py; see here: https://github.com/ultralytics/ultralytics/issues/14662
